ui/raidboss/emulator/data/CombatantState.ts

```diff
@@ -1,76 +1,85 @@
+import { worldNameToWorld } from '../../../../resources/world_id';
 import { PluginCombatantState } from '../../../../types/event';
 
-import Combatant from './Combatant';
+export default class CombatantState implements PluginCombatantState {
+  CurrentWorldID?: number | undefined;
+  WorldID?: number | undefined;
+  WorldName?: string | undefined;
+  BNpcID?: number | undefined;
+  BNpcNameID?: number | undefined;
+  PartyType?: number | undefined;
+  ID?: number | undefined;
+  OwnerID?: number | undefined;
+  WeaponId?: number | undefined;
+  type?: number | undefined;
+  Job?: number | undefined;
+  Level?: number | undefined;
+  Name?: string | undefined;
+  CurrentHP: number;
+  MaxHP: number;
+  CurrentMP: number;
+  MaxMP: number;
+  PosX: number;
+  PosY: number;
+  PosZ: number;
+  Heading: number;
 
-export default class CombatantState {
-  posX: number;
-  posY: number;
-  posZ: number;
-  heading: number;
   targetable: boolean;
-  hp: number;
-  maxHp: number;
-  mp: number;
-  maxMp: number;
 
   // This is a temporary variable used during CombatantTracker initialization and is `delete`d
   // after the combatant states have been determined to keep memory usage low.
   json?: string;
 
-  constructor(
-    posX: number,
-    posY: number,
-    posZ: number,
-    heading: number,
-    targetable: boolean,
-    hp: number,
-    maxHp: number,
-    mp: number,
-    maxMp: number,
-  ) {
-    this.posX = posX;
-    this.posY = posY;
-    this.posZ = posZ;
-    this.heading = heading;
+  constructor(props: Partial<PluginCombatantState>, targetable: boolean) {
+    Object.assign(this, props);
+
+    // Force these values to something sane in case they're not in `props`
```

Comment:
```suggestion
    // Force these values to defaults in case they're not in `props`
```

Should we do this for other fields as well? I guess I don't understand why these get defaulted here but others are `?? 0` elsewhere.

Reply:
These are defaulted here because they're not optional per the `PluginCombatantState` interface itself.

`CombatantState` could instead implement `Partial<PluginCombatantState>`, but that's just kicking this validation responsibility downstream to other locations, leading to duplicate logic.

If other locations are checking these specific values with `?? 0`, that's something that could be cleaned up in those locations.

As for why `PluginCombatantState` has these fields as required, and other fields as optional, that's because I believe that these values were required to always return for `getCombatants` calls. If instead all fields on `PluginCombatantState` should be made required now that it's expected that they should return, then that change can be made, but it's going to make this PR a bit bigger.

Reply:
Ah, ok.  "compatibility with `PluginCombatantState`" is enough of a reason for me, thanks.